### PR TITLE
CSF Plugin command line mode 

### PR DESCRIPTION
**Pull request by DavixDevelop**
Modified [CSF](https://github.com/DavixDevelop/TerraLidar/tree/master/qCSF) plugin for command-line mode usage

### CODE REVIEW

File: plugins/core/Standard/qCSF/README.md [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+
+Available options
+-----------------
+<table>
+	<tr>
+		<th>Command</th>
+		<th>Description</th>
+	</tr>
+	<tr>
+		<td><code>-CSF</code></td>
```
Comment (dgirardeau) [inline plugins/core/Standard/qCSF/README.md:30]: Would it be possible to add this recap in the qCC/CHANGELOG.md file as well?

**Comment by dgirardeau** [comment]
Not sure why the CI is complaining... Do you need to fix something?

**Comment by DavixDevelop** [comment]
> Not sure why the CI is complaining... Do you need to fix something?

I'm looking into it right now. And about the changelog, sure.

File: plugins/core/Standard/qCSF/include/qCSFCommands.h [hunk 2]
```diff
@@ -0,0 +1,270 @@
...
+			csfPC.reserve(count);
+		}
+		catch (const std::bad_alloc&) {
+			return cmd.error("Not enough memory!");
+		}
+
+		for (unsigned i = 0; i < count; i++) {
+			const CCVector3* P = pc->getPoint(i);
+			wl::Point tmpPoint;
+			//tmpPoint.x = P->x;
```
Comment (dgirardeau) [inline plugins/core/Standard/qCSF/include/qCSFCommands.h:83]: to be removed?

File: plugins/core/Standard/qCSF/include/qCSFCommands.h [hunk 3]
```diff
@@ -0,0 +1,270 @@
...
+			//tmpPoint.y = P->y;
+			//tmpPoint.z = P->z;
+			tmpPoint.x = P->x;
+			tmpPoint.y = -P->z;
+			tmpPoint.z = P->y;
+			csfPC.push_back(tmpPoint);
+		}
+
+		//initial parameters
+		static bool csf_postprocessing = false;
```
Comment (dgirardeau) [inline plugins/core/Standard/qCSF/include/qCSFCommands.h:93]: why static?

Comment (dgirardeau) [reply plugins/core/Standard/qCSF/include/qCSFCommands.h:93]: In any case, in this project we set a upper case letters at the beginning of static variable names, and lower case for the others.

File: plugins/core/Standard/qCSF/include/qCSFCommands.h [hunk 4]
```diff
@@ -0,0 +1,270 @@
...
+		csf.params.bSloopSmooth = csf_postprocessing;
+		csf.params.time_step = 0.65;
+		csf.params.class_threshold = class_threshold;
+		csf.params.cloth_resolution = cloth_resolution;
+		csf.params.rigidness = csf_rigidness;
+		csf.params.iterations = MaxIteration;
+
+		std::vector<int> groundIndexes;
+		std::vector<int> offGroundIndexes;
+		ccMesh* clothMesh = nullptr;
```
Comment (dgirardeau) [inline plugins/core/Standard/qCSF/include/qCSFCommands.h:191]: it's a little bit misleading to declare this clothMesh variable here, as well as ExportClothMesh . ExportClothMesh is always false for now, but if it was turned to true, it wouldn't work either (as clothMesh is not managed later).

File: plugins/core/Standard/qCSF/include/qCSFCommands.h [hunk 5]
```diff
@@ -0,0 +1,270 @@
...
+
+struct CommandCSF : public ccCommandLineInterface::Command
+{
+	CommandCSF() : ccCommandLineInterface::Command("CSF", COMMAND_CSF) {}
+
+	virtual bool process(ccCommandLineInterface& cmd) override
+	{
+		cmd.print("[CSF]");
+
+		if (cmd.clouds().size() < 1) {
```
Comment (dgirardeau) [inline plugins/core/Standard/qCSF/include/qCSFCommands.h:63]: In this project we go the line for opened brackets ({)

**Review by dgirardeau — COMMENTED** [review]
Please, add something in the CHANGELOG.md file as well ;)

**Review by dgirardeau — APPROVED** [review]
That's good with me!

**Review by dgirardeau — CHANGES_REQUESTED** [review]
Oops, no, remove the <table> and <tr> from Changelog ;)

**Comment by dgirardeau** [comment]
Sorry about the misunderstanding: I just wanted you to remove the decorations (not the contents!). The contents are super useful...

**Review by dgirardeau — APPROVED** [review]

**Comment by dgirardeau** [comment]
Thanks a lot!

**Comment by ChandrikaMedapati** [comment]
Any Help for -CSF filter from command line. Any Examples to understand more..

**Comment by DavixDevelop** [comment]
> Any Help for -CSF filter from cloud compare command line syntax. Any Examples to understand more.

You can find more in the changelog, under 2.12

**Comment by ChandrikaMedapati** [comment]
I've tried the below command with following versions  through command line CloudCompare_v2.11.2,CloudCompare_v2.11.3,CloudCompare_v2.12.
"C:\Program Files\CloudCompare\CloudCompare.exe" -O -GLOBAL_SHIFT AUTO "C:\Users\cm27280\Downloads\file.laz" -CSF -SCENES SLOPE -CLOTH_RESOLUTION 0.5 -CLASS_THRESHOLD 0.5
But no luck :( 
can anyone please help me in this

**Comment by DavixDevelop** [comment]
> I've tried the below command with following versions  through command line CloudCompare_v2.11.2,CloudCompare_v2.11.3,CloudCompare_v2.12.
> "C:\Program Files\CloudCompare\CloudCompare.exe" -O -GLOBAL_SHIFT AUTO "C:\Users\cm27280\Downloads\file.laz" -CSF -SCENES SLOPE -CLOTH_RESOLUTION 0.5 -CLASS_THRESHOLD 0.5
> But no luck :( 
> can anyone please help me in this
> image
> 

    When you build, did you enable the CSF plugin in cmake? You can test this by opening the built program in checking if the plugin is available under plugins.

**Comment by dgirardeau** [comment]
And mind that only the 2.12.alpha version support this.

**Comment by melissaur1039** [comment]
I keep running into an "Unknown or misplaced command: '-CSF'" error when trying to execute this code that iterates through tiles in a folder, code and error shown in this screenshot:
image
I know the CSF plugin is enabled because it is useable in the CloudCompare program itself. Any ideas? I am brand new to CloudCompare and its command line mode.

**Comment by melissaur1039** [comment]
> Ah that's way too old ;)
> 
> You should try with the latest 2.12 version.

Hey, that worked! Thanks for your help. The school isn't great at keeping all of these programs updated :(

**Comment by samchris3211** [comment]
Hi , i've tried running this in python
subprocess.call(["D:/CloudCompare_v2.12.4_bin_x64/CloudCompare.exe","-AUTO_SAVE","OFF","-C_EXPORT_FMT",file0,"-o", output_file,"-CSF","-SCENES SLOPE","-CLOTH_RESOLUTION","0.5","-CLASS_THRESHOLD","0.5"])

it works fine, however I need to know where the output file is as I can't find it?
The process executes but I just needs it output.

**Comment by dgirardeau** [comment]
Can you maybe use '-SAVE_CLOUDS' or '-SAVE_MESHES' to force the saving of the files?

**Pull request merged**